File: packages/sdk/csharp/packages/admin/AdminAuthClient.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using EdgeBase.Generated;
// EdgeBase C# Admin SDK — AdminAuth
// 서버사이드 유저 관리: CRUD, 커스텀 클레임, 세션 무효화.
// All HTTP calls delegate to Generated Admin Core (GeneratedAdminApi).
// No hardcoded API paths — the core is the single source of truth.

namespace EdgeBase.Admin
{

/// <summary>서버사이드 유저 관리 클라이언트.</summary>
public sealed class AdminAuthClient
{
    private readonly JbHttpClient _http;
    private readonly GeneratedAdminApi _adminCore;

    internal AdminAuthClient(JbHttpClient http)
    {
        _http = http;
        _adminCore = new GeneratedAdminApi(http);
    }

    /// <summary>유저 조회.</summary>
    public async Task<Dictionary<string, object?>> GetUserAsync(
        string userId, CancellationToken ct = default)
    {
        var result = await _adminCore.AdminAuthGetUserAsync(Uri.EscapeDataString(userId), ct);
        return JsonHelper.ExtractNested(result, "user");
    }

    /// <summary>유저 목록 조회 (커서 페이지네이션).</summary>
    public async Task<ListUsersResult> ListUsersAsync(
        int? limit = null, string? cursor = null, CancellationToken ct = default)
    {
        var query = new Dictionary<string, string>();
        if (limit.HasValue) query["limit"] = limit.Value.ToString();
        if (!string.IsNullOrEmpty(cursor)) query["cursor"] = cursor!;

        var result = await _adminCore.AdminAuthListUsersAsync(
            query.Count > 0 ? query : null, ct);
        return new ListUsersResult(
            JsonHelper.ExtractList(result, "users"),
            JsonHelper.ExtractString(result, "cursor"));
    }

    /// <summary>유저 생성 (서버사이드 가입).</summary>
    public async Task<Dictionary<string, object?>> CreateUserAsync(
        string email, string password, string? displayName = null, string? role = null,
        CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?> { ["email"] = email, ["password"] = password };
        if (displayName != null) body["displayName"] = displayName;
        if (role != null) body["role"] = role;

        var result = await _adminCore.AdminAuthCreateUserAsync(body, ct);
        return JsonHelper.ExtractNested(result, "user");
    }

    /// <summary>유저 정보 수정.</summary>
    public async Task<Dictionary<string, object?>> UpdateUserAsync(
        string userId, Dictionary<string, object?> data, CancellationToken ct = default)
    {
        var result = await _adminCore.AdminAuthUpdateUserAsync(
            Uri.EscapeDataString(userId), data, ct);
        return JsonHelper.ExtractNested(result, "user");
    }

    /// <summary>유저 삭제.</summary>
    public Task<Dictionary<string, object?>> DeleteUserAsync(
        string userId, CancellationToken ct = default)
    {
        return _adminCore.AdminAuthDeleteUserAsync(Uri.EscapeDataString(userId), ct);
    }

    /// <summary>커스텀 클레임 설정 (다음 JWT 갱신 시 반영).</summary>
    public Task<Dictionary<string, object?>> SetCustomClaimsAsync(
        string userId, Dictionary<string, object?> claims, CancellationToken ct = default)
    {
        return _adminCore.AdminAuthSetClaimsAsync(
            Uri.EscapeDataString(userId), claims, ct);
    }

    /// <summary>유저의 전체 세션 무효화 (강제 재인증).</summary>
    public Task<Dictionary<string, object?>> RevokeAllSessionsAsync(
        string userId, CancellationToken ct = default)
    {
        return _adminCore.AdminAuthRevokeUserSessionsAsync(
            Uri.EscapeDataString(userId), ct);
    }

    /// <summary>유저의 MFA 비활성화 (관리자 전용, Service Key 필요).</summary>
    public Task<Dictionary<string, object?>> DisableMfaAsync(
        string userId, CancellationToken ct = default)
    {
        return _adminCore.AdminAuthDeleteUserMfaAsync(
            Uri.EscapeDataString(userId), ct);
    }
}

/// <summary>유저 목록 조회 결과.</summary>
public sealed class ListUsersResult
{
    public List<Dictionary<string, object?>> Users { get; }
    public string? Cursor { get; }

    public ListUsersResult(List<Dictionary<string, object?>> users, string? cursor)
    {
        Users  = users;
        Cursor = cursor;
    }
}
}

File: packages/sdk/dart/packages/admin/lib/src/admin_auth_client.dart
// Admin Auth client for server-side user management.
//
// Mirrors JS SDK AdminAuthClient — requires Service Key.
// Only use in server-side (backend Dart) environments.

import 'package:edgebase_core/src/http_client.dart';
import 'package:edgebase_core/src/errors.dart';
import 'generated/admin_api_core.dart';

/// Admin user representation.
class AdminUser {
  final String id;
  final String? email;
  final String? displayName;
  final String? avatarUrl;
  final String? role;
  final String? locale;
  final String? emailVisibility;
  final bool? emailVerified;
  final bool? isAnonymous;
  final bool? disabled;
  final String? createdAt;
  final String? updatedAt;
  final Map<String, dynamic>? metadata;

  AdminUser({
    required this.id,
    this.email,
    this.displayName,
    this.avatarUrl,
    this.role,
    this.locale,
    this.emailVisibility,
    this.emailVerified,
    this.isAnonymous,
    this.disabled,
    this.createdAt,
    this.updatedAt,
    this.metadata,
  });

  factory AdminUser.fromJson(Map<String, dynamic> json) {
    return AdminUser(
      id: json['id'] as String,
      email: json['email'] as String?,
      displayName: json['displayName'] as String?,
      avatarUrl: json['avatarUrl'] as String?,
      role: json['role'] as String?,
      locale: json['locale'] as String?,
      emailVisibility: json['emailVisibility'] as String?,
      emailVerified: json['emailVerified'] == null ? null : (json['emailVerified'] == true || json['emailVerified'] == 1),
      isAnonymous: json['isAnonymous'] == null ? null : (json['isAnonymous'] == true || json['isAnonymous'] == 1),
      disabled: json['disabled'] == null ? null : (json['disabled'] == true || json['disabled'] == 1),
      createdAt: json['createdAt'] as String?,
      updatedAt: json['updatedAt'] as String?,
      metadata: json['metadata'] as Map<String, dynamic>?,
    );
  }
}

/// Options for updating a user via admin API.
class AdminUpdateUserOptions {
  final String? email;
  final String? password;
  final String? displayName;
  final String? avatarUrl;
  final String? role;
  final String? locale;
  final String? emailVisibility;
  final bool? emailVerified;
  final bool? disabled;
  final Map<String, dynamic>? metadata;

  AdminUpdateUserOptions({
    this.email,
    this.password,
    this.displayName,
    this.avatarUrl,
    this.role,
    this.locale,
    this.emailVisibility,
    this.emailVerified,
    this.disabled,
    this.metadata,
  });

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{};
    if (email != null) map['email'] = email;
    if (password != null) map['password'] = password;
    if (displayName != null) map['displayName'] = displayName;
    if (avatarUrl != null) map['avatarUrl'] = avatarUrl;
    if (role != null) map['role'] = role;
    if (locale != null) map['locale'] = locale;
    if (emailVisibility != null) map['emailVisibility'] = emailVisibility;
    if (emailVerified != null) map['emailVerified'] = emailVerified;
    if (disabled != null) map['disabled'] = disabled;
    if (metadata != null) map['metadata'] = metadata;
    return map;
  }
}

/// Admin Auth list result — aligned with server limit/cursor pagination.
class AdminListUsersResult {
  final List<AdminUser> users;
  final String? cursor; // null = no more pages

  AdminListUsersResult({
    required this.users,
    this.cursor,
  });
}

/// Admin Auth client — server-side user management.
class AdminAuthClient {
  final GeneratedAdminApi _core;
  final bool _hasServiceKey;

  AdminAuthClient(HttpClient client, this._hasServiceKey)
      : _core = GeneratedAdminApi(client);

  void _ensureServiceKey() {
    if (!_hasServiceKey) {
      throw EdgeBaseError(
        'AdminAuthClient requires a Service Key. '
        'Initialize EdgeBase with serviceKey option.',
      );
    }
  }

  /// Get a user by ID.
  Future<AdminUser> getUser(String userId) async {
    _ensureServiceKey();
    final resp = await _core.adminAuthGetUser(userId)
        as Map<String, dynamic>;
    // Server wraps response: {"user": {...}}
    final json = resp.containsKey('user')
        ? resp['user'] as Map<String, dynamic>
        : resp;
    return AdminUser.fromJson(json);
  }

  /// List users with cursor-based pagination (aligned with server API).
  Future<AdminListUsersResult> listUsers({
    int limit = 20,
    String? cursor,
  }) async {
    _ensureServiceKey();
    final query = <String, String>{
      'limit': limit.toString(),
    };
    if (cursor != null) query['cursor'] = cursor;

    final json = await _core.adminAuthListUsers(query)
        as Map<String, dynamic>;
    final users = (json['users'] as List<dynamic>)
        .map((e) => AdminUser.fromJson(e as Map<String, dynamic>))
        .toList();
    return AdminListUsersResult(
      users: users,
      cursor: json['cursor'] as String?,
    );
  }

  /// Update a user by ID.
  Future<AdminUser> updateUser(
    String userId,
    AdminUpdateUserOptions data,
  ) async {
    _ensureServiceKey();
    final resp = await _core.adminAuthUpdateUser(userId, data.toJson())
        as Map<String, dynamic>;
    // Server wraps response: {"user": {...}} — unwrap like getUser
    final json = resp.containsKey('user')
        ? resp['user'] as Map<String, dynamic>
        : resp;
    return AdminUser.fromJson(json);
  }

  /// Delete a user by ID.
  Future<void> deleteUser(String userId) async {
    _ensureServiceKey();
    await _core.adminAuthDeleteUser(userId);
  }

  /// Create a user (admin-initiated sign-up).
  Future<AdminUser> createUser({
    required String email,
    required String password,
    String? displayName,
    String? role,
    bool emailVerified = false,
    Map<String, dynamic>? metadata,
  }) async {
    _ensureServiceKey();
    final body = <String, dynamic>{
      'email': email,
      'password': password,
      'emailVerified': emailVerified,
    };
    if (displayName != null) body['displayName'] = displayName;
    if (role != null) body['role'] = role;
    if (metadata != null) body['metadata'] = metadata;

    final resp = await _core.adminAuthCreateUser(body)
        as Map<String, dynamic>;
    // Server wraps response: {"user": {...}, "accessToken": ...}
    final json = resp.containsKey('user')
        ? resp['user'] as Map<String, dynamic>
        : resp;
    return AdminUser.fromJson(json);
  }

  /// Set custom claims for a user (reflected in JWT on next token refresh).
  Future<void> setCustomClaims(
    String userId,
    Map<String, dynamic> claims,
  ) async {
    _ensureServiceKey();
    await _core.adminAuthSetClaims(userId, claims);
  }

  /// Revoke all sessions for a user (force re-authentication).
  Future<void> revokeAllSessions(String userId) async {
    _ensureServiceKey();
    await _core.adminAuthRevokeUserSessions(userId);
  }
}
